File: iosApp/iosApp/ViewModels/SakeShopListViewModel.swift
import Foundation
import Shared
import Combine

@MainActor
class SakeShopListViewModel: ObservableObject {
    @Published var isLoading = true
    @Published var sakeShops: [SakeShop] = []
    @Published var errorMessage: String?
    
    private let getSakeShopsUseCase: GetSakeShopsUseCase
    
    init() {
        self.getSakeShopsUseCase = DependencyContainer.shared.getSakeShopsUseCase()
        loadSakeShops()
    }
    
    func loadSakeShops() {
        isLoading = true
        errorMessage = nil
        
        Task {
            do {
                let shops = try await getSakeShopsUseCase.invoke()
                await MainActor.run {
                    self.sakeShops = shops
                    self.isLoading = false
                }
            } catch {
                await MainActor.run {
                    self.errorMessage = error.localizedDescription
                    self.isLoading = false
                }
            }
        }
    }
}

File: composeApp/src/androidMain/kotlin/com/bimm/takehomeassignmnent/presentation/viewmodel/SakeShopListViewModel.kt
package com.bimm.takehomeassignmnent.presentation.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.bimm.takehomeassignmnent.domain.model.SakeShop
import com.bimm.takehomeassignmnent.domain.usecase.GetSakeShopsUseCase
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

data class SakeShopListUiState(
    val isLoading: Boolean = true,
    val sakeShops: List<SakeShop> = emptyList(),
    val errorMessage: String? = null
)

class SakeShopListViewModel(
    private val getSakeShopsUseCase: GetSakeShopsUseCase
) : ViewModel() {
    
    private val _uiState = MutableStateFlow(SakeShopListUiState())
    val uiState: StateFlow<SakeShopListUiState> = _uiState.asStateFlow()
    
    init {
        loadSakeShops()
    }
    
    fun loadSakeShops() {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, errorMessage = null)
            
            getSakeShopsUseCase()
                .onSuccess { shops ->
                    _uiState.value = _uiState.value.copy(
                        isLoading = false,
                        sakeShops = shops
                    )
                }
                .onFailure { exception ->
                    _uiState.value = _uiState.value.copy(
                        isLoading = false,
                        errorMessage = exception.message ?: "Unknown error occurred"
                    )
                }
        }
    }
}